### ISSUE TITLE
From URI does not match AoR

### ISSUE THREAD

**Issue by juha-h** [baresip/baresip#2576]
My account is like this:
```
--- sip:test%401@foo.bar ---
 nrefs:     1
 cuser:     test%401-0x55ab26979e70
 pub-gruu:  
 Supported: gruu,path,outbound,replaces,norefersub

Account:
 address:      <sip:test%401@foo.bar>;audio_codecs=PCMU,PCMA;regint=0;pubint=0;...
 luri:         sip:test%401@foo.bar
 aor:          sip:test%401@foo.bar
```
When the UA makes a call, From header looks like this:
```
INVITE sip:bar@test.tutpro.com SIP/2.0
Via: SIP/2.0/TCP 192.168.188.215:5020;branch=z9hG4bK986088d8c51fb5b6;rport
Contact: <sip:test%401-0x55ab26979e70@192.168.188.215:5020;transport=tcp>
Max-Forwards: 70
To: <sip:bar@test.tutpro.com>
From: <sip:test%25401@foo.bar>;tag=e0f802f01da481de
```
The AoR has character @ that according to RFC 3261 needs to be escaped and written as %40. 

Why is the escape prefix % escaped?

**Comment by maximilianfridrich** [comment]
I added a few commits regarding URI escaping. The account URI you configure is taken literally and then escaped, so I guess you have to put it in the actual "@" sign.

Sorry about the confusion and the change in behavior. Let me me know if you see any issues in this regard.

**Comment by juha-h** [comment]
Then I tried with AoR `sip:test@1@foo.bar` and From URI was
```
From: <sip:test@1@foo.bar>;tag=0860ddae856308f8
```
So looks like baresip does not check AoR validity.

**Comment by juha-h** [comment]
@maximilianfridrich Which commits you refer to?

**Comment by maximilianfridrich** [comment]
Ah yes, @ signs are a bit special because they are not properly handled yet in uri_decode.

https://github.com/baresip/re/pull/740
https://github.com/baresip/re/pull/762
https://github.com/baresip/re/pull/770

**Comment by juha-h** [comment]
In my opinion, barasip must allow using any AoR that are valid according to RFC3261 syntax and it should check that the AoR user has written in accounts file is valid when baresip starts.

**Comment by maximilianfridrich** [comment]
Yes, ideally. The @ sign is tripping it up. Except for that, everything is else is fine and you can put any AOR in there and baresip/re will properly escape.

We allow our users to create AORs and we can not expect them to escape the strings, this must be handled by the application. They put in UTF-8.

**Comment by juha-h** [comment]
Yes, application can do and acept anything, but when the app creates accounts file, AoR in that file must be written according to RFC3261 syntax.  And then when baresip starts it should check validity of the AoR.

**Comment by juha-h** [comment]
I'll label this issue as a bug.  It was noticed by a user of my Android app.

**Comment by maximilianfridrich** [comment]
So your application does the escaping and writes the escaped string?

I guess we use it differently, we write the "raw" string and let BareSIP/re do the escaping.

**Comment by juha-h** [comment]
Somehow I missed those three commits.  Last one (770) has this explanation:

`The SIP URI we receive in the Refer-To header is already escaped, so before triggering UA_EVENT_REFER, we unescape it. Otherwise, we would double-escape this URI erroneously when sending the INVITE request to this URI.`

In my opinion, baresip core should not do any (un)escaping whatsoever.  All (un)escaping should be done only by the application.

In case of REFER, if baresip handles it without consulting the user, there is no problem with the INVITE.  If URI is escaped in REFER, it will stay escaped in INVITE.  If app consults the user, app unescapes the URI when it shows it to the user and then escapes it again if INVITE will be sent.

**Comment by juha-h** [comment]
baresip and re libs should not do any application level stuff, but just work according to RFC3261.  Otherwise baresip/re are usable only for some applications.

**Comment by juha-h** [comment]
So when I give `u` menu command, baresip should never show an URI that is invalid according to RFC3261, but reject the account at start.  So not like this:
```
address:      <sip:first last@foo.bar>;";ptime=20;audio_codecs=PCMU,PCMA
 luri:         sip:first last@foo.bar
```
luri is not a valid URI and this should not happen.

**Comment by juha-h** [comment]
RFC3261 allows also non-printable characters in URI userpart as long as they are escaped.  Now when account's AoR user % character is escaped, there is no way to use such characters.

**Comment by juha-h** [comment]
So please revert those commits.  baresip/re should remain basic RFC3261 compliant libraries, not tailored to some specific application usages.

**Comment by maximilianfridrich** [comment]
Again, we interpreted the text in the accounts file or the string received by the command `uanew` as raw, unescaped strings. Your interpretation is probably valid too, and maybe would be better/easier.

Keep in mind though that this means you application needs to know the RFC quite well too, and the BNF is not trivial. Further, it needs to know exactly how the URIs are assembled and sent if there are different fields for user, host, display name etc - e.g. if the display name is sent with double quotes around it or not. See:

```
      UTF8-NONASCII   =  %xC0-DF 1UTF8-CONT
                      /  %xE0-EF 2UTF8-CONT
                      /  %xF0-F7 3UTF8-CONT
                      /  %xF8-Fb 4UTF8-CONT
                      /  %xFC-FD 5UTF8-CONT
      UTF8-CONT       =  %x80-BF

      quoted-string  =  SWS DQUOTE *(qdtext / quoted-pair ) DQUOTE
      qdtext         =  LWS / %x21 / %x23-5B / %x5D-7E
                        / UTF8-NONASCII
      quoted-pair  =  "\" (%x00-09 / %x0B-0C
                      / %x0E-7F)

      display-name   =  *(token LWS)/ quoted-string

      name-addr      =  [ display-name ] LAQUOT addr-spec RAQUOT
      addr-spec      =  SIP-URI / SIPS-URI / absoluteURI
      display-name   =  *(token LWS)/ quoted-string
```

I am not sure yet what the correct interpretation of the accounts file and uanew argument is. Could someone else chime in? @alfredh @sreimers

**Comment by maximilianfridrich** [comment]
For reference, here is how PJSIP goes about it: https://docs.pjsip.org/en/latest/specific-guides/sip/uri_escaping.html

It uses both: pjsip_sip_uri structs are always unescaped, while URIs as strings are always escaped.

**Comment by juha-h** [comment]
@cspiel1 I have no problem with menu `uanew` command. Menu  is baresip application and can do whatever is convenient for the user.  But baresip/re libs should be RC3261 compliant.  Now they are not and there is no way I can create an account in accounts file that has escaped characters in AoR userpart, which is perfectly legal according to RFC3261.
```

**Comment by juha-h** [comment]
I'm against a config option to turn on (un)escaping in baresip/re libs.  The libs should be kept simple and RFC3261 compliant.  As I already wrote, regarding REFER, for example, there is no need for (un)escaping in the libs.

**Comment by maximilianfridrich** [comment]
We are still RFC compliant. Now, the special characters in your accounts file are escaped for you. Except for an @ in the user part which is a known limitation that BareSIP has never been able to cope with due to the `uri_decode` implementation.

**Comment by juha-h** [comment]
So you can do whatever is needed `************ here` in `menu.c/create_ua():
```
	if (str_isset(carg->prm)) {
                (void)re_hprintf(pf, "Creating UA for %s ...\n", carg->prm);
                ************ here
		err = ua_alloc(&ua, carg->prm);
		if (err)
			goto out;
	}
```
so that ua_alloc() receives RFC3261 compliant AoR.

**Comment by juha-h** [comment]
For it isn't RFC3261 compliant.  I cannot use an AoR that has

`escaped = "%" HEXDIG HEXDIG`

characters in AoR.

I cannot understand why you can't write your app so that there is no need to break baresip/re lib.

**Comment by maximilianfridrich** [comment]
How about this: We remove the escaping from the core as you suggested and move the escaping to the accounts module.

Not sure if after that BareSIP handles all URIs correctly, since URIs in From, To, Route,... headers come from different places and it looks like not everyone was on the same page if these are passed around escaped or unescaped.

**Comment by juha-h** [comment]
Regarding `struct uri` `user` field, best to have its characters escaped, because they are always escaped anyway when SIP request is received or sent.  If there is need for (un)escaping, that can be done at the user interface.

**Comment by cspiel1** [comment]
Then we have a problem. We want to have UTF-8 accounts file. Because don't want to have SIP escaping code in our application. Same for the `uanew` command of module `menu`.

Maybe some automatic detection is possible.

**Comment by maximilianfridrich** [comment]
@juha-h We implemented the escaping exactly **because we want to be RFC compliant**. The only issue we are having is whether the input into BareSIP like creating accounts, contacts, etc. is already escaped or not. We will always be RFC compliant.

This question needs to be answered: Does BareSIP expect escaped input or unescaped input?

**Comment by juha-h** [comment]
> @juha-h We implemented the escaping exactly **because we want to be RFC compliant**. The only issue we are having is whether the input into BareSIP like creating accounts, contacts, etc. is already escaped or not.
> 
> This question needs to be answered: Does BareSIP expect escaped input or unescaped input?

Please show BNF syntax of your AoR `user` alternative.

**Comment by maximilianfridrich** [comment]
@juha-h I had another discussion with @cspiel1, we agreed to revert my recent URI escaping commits to restore the old behavior. We will make sure to enter escaped URIs into BareSIP.

I apologize for causing breaks, we were not fully aware of how others used BareSIP.

**Comment by juha-h** [comment]
Maximilian Fridrich writes:

> @juha-h I had another discussion with @cspiel1, we agreed to revert my
> recent URI escaping commits to restore the old behavior. We will make
> sure to enter escaped URIs into BareSIP.

OK, thanks.
